src/lib/runtime/files/hostname/hostname.c

```diff
@@ -0,0 +1,70 @@
+/* 
+ * Copyright (c) 2017, SingularityWare, LLC. All rights reserved.
+ * 
+ * This software is licensed under a 3-clause BSD license.  Please
+ * consult LICENSE file distributed with the sources of this project regarding
+ * your rights to use or distribute this software.
+ * 
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <limits.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <grp.h>
+#include <pwd.h>
+
+#include "config.h"
+#include "util/file.h"
+#include "util/util.h"
+#include "util/registry.h"
+#include "util/config_parser.h"
+#include "util/message.h"
+#include "util/privilege.h"
+
+#include "../file-bind.h"
+#include "../../runtime.h"
+
+
+int _singularity_runtime_files_hostname(void) {
+    FILE *hostname_fd;
+    char *tmpdir = singularity_registry_get("SESSIONDIR");
+    char *hostname_file = "/etc/hostname";
+    char *containerdir = CONTAINER_FINALDIR;
+    char *source = joinpath(tmpdir, "/hostname");
+    char *target = joinpath(containerdir, hostname_file);
+    char *hostname = singularity_registry_get("HOSTNAME");
+
+    if ( hostname == NULL ) {
+        singularity_message(DEBUG, "Setting container hostname not requested by user\n");
+        return(0);
+    }
+
+    singularity_message(DEBUG, "Check if /etc/hostname is present in container\n");
+    if ( is_file(target) < 0 ) {
+        singularity_message(VERBOSE, "/etc/hostname doesn't exists, skipping\n");
+        return(0);
+    }
+
+    hostname_fd = fopen(source, "w+");
+    if ( hostname_fd == NULL ) {
+        singularity_message(ERROR, "Couldn't create hostname session file\n");
+        ABORT(255);
+    }
+
+    if ( strlen(hostname) > HOST_NAME_MAX ) {
+        hostname[HOST_NAME_MAX] = '\0';
+    }
+
+    fprintf(hostname_fd, "%s\n", hostname);
```

Comment:
I am pretty sure that we need to set the hostname directly via systemcall, not via file and count on systemd/SysVinit/upstart/pick_your_favorite_init or distribution to do it all the same.

Reply:
using only syscall won't work when --boot option is given, init systems set hostname based on /etc/hostname content (container default is often localhost.localdomain), so both are required to handle this case